<a href="https://colab.research.google.com/github/PhilR1127/Data-601-Final-Project-/blob/main/D699V11_1_23_Copy_of_BigQuery_bquxjob_4aaf7004_18b86b0b98c.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# @title Setup
from google.colab import auth
from google.cloud import bigquery
from google.colab import data_table

project = 'data-699' # Project ID inserted based on the query results selected to explore
location = 'US' # Location inserted based on the query results selected to explore
client = bigquery.Client(project=project, location=location)
data_table.enable_dataframe_formatter()
auth.authenticate_user()

## Reference SQL syntax from the original job
Use the ```jobs.query```
[method](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query) to
return the SQL syntax from the job. This can be copied from the output cell
below to edit the query now or in the future. Alternatively, you can use
[this link](https://console.cloud.google.com/bigquery?j=data-699:US:bquxjob_4aaf7004_18b86b0b98c)
back to BigQuery to edit the query within the BigQuery user interface.

In [4]:
# Running this code will display the query used to generate your previous job

job = client.get_job('bquxjob_4aaf7004_18b86b0b98c') # Job ID inserted based on the query results selected to explore
print(job.query)

SELECT subject_id,AGE_patient,AVG_HR, AVG_SBP, AVG_DBP,AVG_O2SAT,atime,dtime,inptdeath,Date_of_Death FROM `data-699.12345.10_30_Patient_Admissions_for_Modeling` 


# Result set loaded from BigQuery job as a DataFrame
Query results are referenced from the Job ID ran from BigQuery and the query
does not need to be re-run to explore results. The ```to_dataframe```
[method](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJob.html#google.cloud.bigquery.job.QueryJob.to_dataframe)
downloads the results to a Pandas DataFrame by using the BigQuery Storage API.

To edit query syntax, you can do so from the BigQuery SQL editor or in the
```Optional:``` sections below.

In [5]:
# Running this code will read results from your previous job

job = client.get_job('bquxjob_4aaf7004_18b86b0b98c') # Job ID inserted based on the query results selected to explore
results = job.to_dataframe()
results

subject_id  AGE_patient        AVG_HR     AVG_SBP    AVG_DBP  \
0         18179783           86  75.377358491  135.452830  68.943396   
1         11756261           62  65.842105263  139.055556  68.611111   
2         16477638           61  75.705882353  147.745098  67.725490   
3         13774741           91  86.903225806  138.774194  74.677419   
4         12844772           58  68.500000000  142.833333  77.000000   
...            ...          ...           ...         ...        ...   
119352    17027269           18  62.500000000  131.500000  88.000000   
119353    17946600           18  92.000000000  126.000000  75.000000   
119354    19531581           18  95.000000000  118.600000  80.200000   
119355    17251355           18  70.000000000  109.000000  74.666667   
119356    14332783           18  74.000000000  103.500000  62.250000   

            AVG_O2SAT               atime               dtime  inptdeath  \
0        98.340909091 2208-04-30 16:47:00 2208-05-03 14:22:00          0   
1        98.842105263 2201-10-12 08:45:00 2201-10-14 18:06:00          0   
2        98.234042553 2210-10-09 14:33:00 2210-10-11 15:05:00          0   
3        98.322580645 2210-10-26 20:21:00 2210-10-30 12:30:00          0   
4        96.666666667 2211-01-14 13:14:00 2211-01-17 12:34:00          1   
...               ...                 ...                 ...        ...   
119352   98.000000000 2189-10-04 17:28:00 2189-10-12 18:00:00          0   
119353   98.000000000 2189-07-19 21:00:00 2189-07-26 14:30:00          0   
119354   99.600000000 2196-08-17 18:40:00 2196-08-20 11:56:00          0   
119355   99.666666667 2193-05-14 06:28:00 2193-05-16 11:15:00          0   
119356  100.000000000 2201-05-15 20:10:00 2201-05-18 15:10:00          0   

       Date_of_Death  
0         2212-01-22  
1         2212-01-14  
2         2211-09-16  
3         2211-02-27  
4         2211-01-17  
...              ...  
119352           NaT  
119353           NaT  
119354           NaT  
119355           NaT  
119356           NaT  

[119357 rows x 10 columns]

## Show descriptive statistics using describe()
Use the ```pandas DataFrame.describe()```
[method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html)
to generate descriptive statistics. Descriptive statistics include those that
summarize the central tendency, dispersion and shape of a dataset’s
distribution, excluding ```NaN``` values. You may also use other Python methods
to interact with your data.

In [37]:
results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119357 entries, 0 to 119356
Data columns (total 15 columns):
 #   Column                    Non-Null Count   Dtype          
---  ------                    --------------   -----          
 0   subject_id                119357 non-null  Int64          
 1   AGE_patient               119357 non-null  Int64          
 2   AVG_HR                    118846 non-null  float64        
 3   AVG_SBP                   118825 non-null  float64        
 4   AVG_DBP                   118825 non-null  float64        
 5   AVG_O2SAT                 117452 non-null  float64        
 6   atime                     119357 non-null  datetime64[ns] 
 7   dtime                     119357 non-null  datetime64[ns] 
 8   inptdeath                 119357 non-null  Int64          
 9   Date_of_Death             17676 non-null   datetime64[ns] 
 10  Pulse_Pressure            118825 non-null  float64        
 11  Max_HR                    119357 non-null  Int64    

In [36]:
results.describe()

subject_id  AGE_patient         AVG_HR        AVG_SBP  \
count         119357.0     119357.0  118846.000000  118825.000000   
mean   15013355.599244    54.879974      80.746417     128.253461   
std     2888151.019781    20.220418      14.192993      17.299130   
min         10000032.0         18.0       7.000000      10.000000   
25%         12502529.0         38.0      71.000000     116.076923   
50%         15026826.0         56.0      79.500000     126.800000   
75%         17521183.0         71.0      89.071429     138.666667   
max         19999987.0         91.0     197.909091     293.000000   

             AVG_DBP      AVG_O2SAT  inptdeath  Pulse_Pressure      Max_HR  \
count  118825.000000  117452.000000   119357.0   118825.000000    119357.0   
mean       74.471090      97.913155   0.039646      110.326004  153.120026   
std        93.537499       2.792009   0.195126       33.766620   20.220418   
min         6.000000       0.000000        0.0       10.666667       117.0   
25%        65.777778      97.000000        0.0      100.333333       137.0   
50%        72.666667      98.166667        0.0      109.083333       152.0   
75%        80.000000      99.000000        0.0      118.488889       170.0   
max     26028.500000     499.800000        1.0     8745.166667       190.0   

       Est_vO2_Max             Length_of_Stay     Time_to_Outpatient_Death  
count     119357.0                     119357                        17676  
mean          <NA>  4 days 05:36:44.826193687   98 days 08:18:55.037338765  
std           <NA>  6 days 14:38:54.676416964  216 days 18:41:20.077168744  
min           <NA>          -1 days +01:21:00         -2172 days +06:30:00  
25%           <NA>            0 days 22:48:00            -1 days +23:29:30  
50%           <NA>            2 days 12:19:00             26 days 11:00:00  
75%           <NA>            4 days 20:52:00            132 days 06:41:00  
max           <NA>          234 days 01:00:00           3901 days 20:30:00

Need to add new variables and transform exisitng ones:

In [59]:
import pandas as pd
import numpy as np


In [60]:
results['AVG_HR'] = pd.to_numeric(results['AVG_HR'])

In [61]:
results['AVG_O2SAT'] = pd.to_numeric(results['AVG_O2SAT'])

Create three new variables:
1. Pulse Pressure,
2.Maximum Heart Rate, and
3.Estimated Vo2 Max

In [62]:
results['Pulse_Pressure'] = (((2*results['AVG_SBP']))+(results['AVG_DBP']))/3

In [63]:
results['Max_HR'] = (208-(results['AGE_patient']))

In [64]:
results['Est_vO2_Max'] = (results['Max_HR'])/(results['AVG_HR'])*15.3

Create three additional time related variables:
1. Length of Stay
2.Time to Patient Death post Discharge

In [65]:
results['Length_of_Stay']=((results['dtime'])-(results['atime']))

In [66]:
results['Date_of_Death']= pd.to_datetime(results['Date_of_Death'])

In [67]:
results['Time_to_Outpatient_Death'] = ((results['Date_of_Death'])-results['dtime'])

Now I want to explore the new table with the new variables

In [68]:
results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119357 entries, 0 to 119356
Data columns (total 15 columns):
 #   Column                    Non-Null Count   Dtype          
---  ------                    --------------   -----          
 0   subject_id                119357 non-null  Int64          
 1   AGE_patient               119357 non-null  Int64          
 2   AVG_HR                    118846 non-null  float64        
 3   AVG_SBP                   118825 non-null  float64        
 4   AVG_DBP                   118825 non-null  float64        
 5   AVG_O2SAT                 117452 non-null  float64        
 6   atime                     119357 non-null  datetime64[ns] 
 7   dtime                     119357 non-null  datetime64[ns] 
 8   inptdeath                 119357 non-null  Int64          
 9   Date_of_Death             17676 non-null   datetime64[ns] 
 10  Pulse_Pressure            118825 non-null  float64        
 11  Max_HR                    119357 non-null  Int64    

In [48]:
from matplotlib import pyplot
from sklearn.linear_model import LogisticRegression
from sklearn.utils import shuffle

In [69]:
results.nunique()

subject_id                  119357
AGE_patient                     73
AVG_HR                       15245
AVG_SBP                      23543
AVG_DBP                      20249
AVG_O2SAT                     5227
atime                       118936
dtime                       118644
inptdeath                        2
Date_of_Death                13641
Pulse_Pressure               30567
Max_HR                          73
Est_vO2_Max                  64381
Length_of_Stay               21578
Time_to_Outpatient_Death     12486
dtype: int64

In [70]:
data_table.disable_dataframe_formatter()

In [71]:
results.head(8)

subject_id  AGE_patient     AVG_HR     AVG_SBP    AVG_DBP  AVG_O2SAT  \
0    18179783           86  75.377358  135.452830  68.943396  98.340909   
1    11756261           62  65.842105  139.055556  68.611111  98.842105   
2    16477638           61  75.705882  147.745098  67.725490  98.234043   
3    13774741           91  86.903226  138.774194  74.677419  98.322581   
4    12844772           58  68.500000  142.833333  77.000000  96.666667   
5    15338454           67  75.684211  153.842105  69.894737  96.562500   
6    12648465           49  82.193833  108.553648  69.137339  99.378378   
7    15566010           86  74.393939  104.878788  51.439394  97.424242   

                atime               dtime  inptdeath Date_of_Death  \
0 2208-04-30 16:47:00 2208-05-03 14:22:00          0    2212-01-22   
1 2201-10-12 08:45:00 2201-10-14 18:06:00          0    2212-01-14   
2 2210-10-09 14:33:00 2210-10-11 15:05:00          0    2211-09-16   
3 2210-10-26 20:21:00 2210-10-30 12:30:00          0    2211-02-27   
4 2211-01-14 13:14:00 2211-01-17 12:34:00          1    2211-01-17   
5 2210-12-01 17:55:00 2210-12-04 02:15:00          1    2210-12-04   
6 2210-07-20 03:56:00 2210-09-16 02:00:00          1    2210-09-16   
7 2210-09-07 20:50:00 2210-09-11 14:25:00          1    2210-09-11   

   Pulse_Pressure  Max_HR  Est_vO2_Max   Length_of_Stay  \
0      113.283019     122    24.763404  2 days 21:35:00   
1      115.574074     146    33.926619  2 days 09:21:00   
2      121.071895     147    29.708392  2 days 00:32:00   
3      117.408602     117    20.598775  3 days 16:09:00   
4      120.888889     150     33.50365  2 days 23:20:00   
5      125.859649     141    28.503964  2 days 08:20:00   
6       95.414878     159    29.597111 57 days 22:04:00   
7       87.065657     122    25.090754  3 days 17:35:00   

  Time_to_Outpatient_Death  
0       1358 days 09:38:00  
1       3743 days 05:54:00  
2        339 days 08:55:00  
3        119 days 11:30:00  
4        -1 days +11:26:00  
5        -1 days +21:45:00  
6        -1 days +22:00:00  
7        -1 days +09:35:00

In [83]:
results.describe()

subject_id  AGE_patient         AVG_HR        AVG_SBP  \
count         119357.0     119357.0  118846.000000  118825.000000   
mean   15013355.599244    54.879974      80.746417     128.253461   
std     2888151.019781    20.220418      14.192993      17.299130   
min         10000032.0         18.0       7.000000      10.000000   
25%         12502529.0         38.0      71.000000     116.076923   
50%         15026826.0         56.0      79.500000     126.800000   
75%         17521183.0         71.0      89.071429     138.666667   
max         19999987.0         91.0     197.909091     293.000000   

             AVG_DBP      AVG_O2SAT  inptdeath  Pulse_Pressure      Max_HR  \
count  118825.000000  117452.000000   119357.0   118825.000000    119357.0   
mean       74.471090      97.913155   0.039646      110.326004  153.120026   
std        93.537499       2.792009   0.195126       33.766620   20.220418   
min         6.000000       0.000000        0.0       10.666667       117.0   
25%        65.777778      97.000000        0.0      100.333333       137.0   
50%        72.666667      98.166667        0.0      109.083333       152.0   
75%        80.000000      99.000000        0.0      118.488889       170.0   
max     26028.500000     499.800000        1.0     8745.166667       190.0   

         Est_vO2_Max             Length_of_Stay     Time_to_Outpatient_Death  
count  118846.000000                     119357                        17676  
mean       29.849081  4 days 05:36:44.826193687   98 days 08:18:55.037338765  
std         6.414420  6 days 14:38:54.676416964  216 days 18:41:20.077168744  
min        11.444400          -1 days +01:21:00         -2172 days +06:30:00  
25%        25.442697            0 days 22:48:00            -1 days +23:29:30  
50%        29.327688            2 days 12:19:00             26 days 11:00:00  
75%        33.636641            4 days 20:52:00            132 days 06:41:00  
max       340.971429          234 days 01:00:00           3901 days 20:30:00

I realized I had somehow created the new Est_vO2_Max column as a "Float64" (capital F) dtype instead of a "float64" dtype .. not sure how, but found a reference to how to fix this issue on StackOverflow...https://stackoverflow.com/questions/69201668/pandas-float64-vs-float64-dtypes-note-capitalization-causing-non-numeric-error not sure what the original issue was though

In [85]:
results['Est_vO2_Max'] = results['Est_vO2_Max'].values.astype(float)

In [86]:
results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119357 entries, 0 to 119356
Data columns (total 15 columns):
 #   Column                    Non-Null Count   Dtype          
---  ------                    --------------   -----          
 0   subject_id                119357 non-null  Int64          
 1   AGE_patient               119357 non-null  Int64          
 2   AVG_HR                    118846 non-null  float64        
 3   AVG_SBP                   118825 non-null  float64        
 4   AVG_DBP                   118825 non-null  float64        
 5   AVG_O2SAT                 117452 non-null  float64        
 6   atime                     119357 non-null  datetime64[ns] 
 7   dtime                     119357 non-null  datetime64[ns] 
 8   inptdeath                 119357 non-null  Int64          
 9   Date_of_Death             17676 non-null   datetime64[ns] 
 10  Pulse_Pressure            118825 non-null  float64        
 11  Max_HR                    119357 non-null  Int64    

In [87]:
results.describe()

subject_id  AGE_patient         AVG_HR        AVG_SBP  \
count         119357.0     119357.0  118846.000000  118825.000000   
mean   15013355.599244    54.879974      80.746417     128.253461   
std     2888151.019781    20.220418      14.192993      17.299130   
min         10000032.0         18.0       7.000000      10.000000   
25%         12502529.0         38.0      71.000000     116.076923   
50%         15026826.0         56.0      79.500000     126.800000   
75%         17521183.0         71.0      89.071429     138.666667   
max         19999987.0         91.0     197.909091     293.000000   

             AVG_DBP      AVG_O2SAT  inptdeath  Pulse_Pressure      Max_HR  \
count  118825.000000  117452.000000   119357.0   118825.000000    119357.0   
mean       74.471090      97.913155   0.039646      110.326004  153.120026   
std        93.537499       2.792009   0.195126       33.766620   20.220418   
min         6.000000       0.000000        0.0       10.666667       117.0   
25%        65.777778      97.000000        0.0      100.333333       137.0   
50%        72.666667      98.166667        0.0      109.083333       152.0   
75%        80.000000      99.000000        0.0      118.488889       170.0   
max     26028.500000     499.800000        1.0     8745.166667       190.0   

         Est_vO2_Max             Length_of_Stay     Time_to_Outpatient_Death  
count  118846.000000                     119357                        17676  
mean       29.849081  4 days 05:36:44.826193687   98 days 08:18:55.037338765  
std         6.414420  6 days 14:38:54.676416964  216 days 18:41:20.077168744  
min        11.444400          -1 days +01:21:00         -2172 days +06:30:00  
25%        25.442697            0 days 22:48:00            -1 days +23:29:30  
50%        29.327688            2 days 12:19:00             26 days 11:00:00  
75%        33.636641            4 days 20:52:00            132 days 06:41:00  
max       340.971429          234 days 01:00:00           3901 days 20:30:00

In [88]:
from pandas.core.api import notnull
results.notnull().sum()

subject_id                  119357
AGE_patient                 119357
AVG_HR                      118846
AVG_SBP                     118825
AVG_DBP                     118825
AVG_O2SAT                   117452
atime                       119357
dtime                       119357
inptdeath                   119357
Date_of_Death                17676
Pulse_Pressure              118825
Max_HR                      119357
Est_vO2_Max                 118846
Length_of_Stay              119357
Time_to_Outpatient_Death     17676
dtype: int64

In [89]:
results.isnull().sum()

subject_id                       0
AGE_patient                      0
AVG_HR                         511
AVG_SBP                        532
AVG_DBP                        532
AVG_O2SAT                     1905
atime                            0
dtime                            0
inptdeath                        0
Date_of_Death               101681
Pulse_Pressure                 532
Max_HR                           0
Est_vO2_Max                    511
Length_of_Stay                   0
Time_to_Outpatient_Death    101681
dtype: int64

Comparing "isnull_sum" variable counts to "notnull.sum() varaible counts shows that I will need to:

1- either remove or normalize the small number of Null values in the AVG_HR,
AVG_SBP, AVG_DBP, AVG_O2SAT, and AVG_Pulse_Pressure

2 - Set "Date of Death" to the dependant varaible for logisitic regression. There are 17,676 NOT-Null values in this column... these would correspond to a death event.. either inpatient or outpatient (Set "death" = 1)
There are 101,681 Null values in this column ... these would correspond to a patient living one year post discharge (Set "living" = 0)